lib/pipeline.ts

```diff
@@ -255,25 +259,94 @@ export class Pipeline extends cdk.Construct {
     }));
   }
 
-  private addFailureAlarm(title?: string) {
-    const pipelineFailureTopic = new sns.Topic(this, 'PipelineFailureTopic');
+  private addFailureAlarm(title?: string): cloudwatch.Alarm {
+    return new PipelineWatcher(this, 'PipelineWatcher', {
+      pipeline: this.pipeline,
+      title
+    }).alarm;
+  }
+}
+
+interface PipelineWatcherProps {
+  /**
+   * Code Pipeline to monitor for failed stages
+   */
+  pipeline: cpipeline.Pipeline;
+
+  /**
+   * Set the pipelineName of the alarm description.
+   *
+   * Description is set to 'Pipeline <title> has failed stages'
+   *
+   * @default pipeline's name
+   */
+  title?: string;
+}
+
+/**
+ * Construct which watches a Code Pipeline for failed stages and raises an alarm
+ * if there are any failed stages.
+ *
+ * A function runs every minute and calls GetPipelineState for the provided pipeline's
+ * name, counts the number of failed stages and emits a JSON log { failedCount: <number> }.
+ * A metric filter is then configured to track this value as a CloudWatch metric, and
+ * a corresponding alarm is set to fire when the maximim value of a single 5-minute interval
+ * is >= 1.
+ */
+class PipelineWatcher extends cdk.Construct {
+  public readonly alarm: cloudwatch.Alarm;
+
+  constructor(parent: cdk.Construct, name: string, props: PipelineWatcherProps) {
+    super(parent, name);
+
+    const pipelineWatcher = new lambda.Function(this, 'Poller', {
+      handler: 'index.handler',
+      runtime: lambda.Runtime.NodeJS810,
+      code: lambda.Code.inline(fs.readFileSync(path.join(__dirname, 'pipeline-watcher.js')).toString('utf8')),
+      environment: {
+        pipelineName: props.pipeline.pipelineName
+      }
+    });
+
+    // See https://github.com/awslabs/aws-cdk/issues/1340 for exposing grants on the pipeline.
+    pipelineWatcher.addToRolePolicy(new iam.PolicyStatement()
+      .addResource(props.pipeline.pipelineArn)
+      .addAction('codepipeline:GetPipelineState'));
+
+    // Explicitly and pre-emptively create the function's log group because we need it for the metric filter.
+    const logGroup = new logs.LogGroup(this, 'Logs', {
+      logGroupName: new cdk.FnConcat('/aws/lambda/', pipelineWatcher.functionName).toString(),
```

Comment:
We prefer not to use `FnConcat` explicitly and let the stringified token magic do it's trick:

    `/aws/lambda/${pipelineWatcher.functionName}`

Reply:
Sweet!